Add instrumentation for thrift (#18405)
open-telemetry/opentelemetry-java-instrumentation: instrumentation/thrift-0.13/library/README.md

> # Library Instrumentation for gRPC 0.13.0+

Reviewer: [General] Title says gRPC — copy/paste from the gRPC README.

```suggestion
# Library Instrumentation for Apache Thrift 0.13.0+
```

Provides OpenTelemetry instrumentation for [Apache Thrift](https://thrift.apache.org/).

## Quickstart

### Add the following dependencies to your project

> Replace `OPENTELEMETRY_VERSION` with the [latest release]( https://central.sonatype.com/artifact/io.opentelemetry.instrumentation/opentelemetry-grpc-1.6).

Reviewer: [General] Maven Central link points to the gRPC artifact instead of Thrift.

```suggestion
Replace `OPENTELEMETRY_VERSION` with the [latest release](https://central.sonatype.com/artifact/io.opentelemetry.instrumentation/opentelemetry-thrift-0.13).
```

For Maven, add the following to your `pom.xml` dependencies:

```xml
<dependencies>
  <dependency>
    <groupId>io.opentelemetry.instrumentation</groupId>
    <artifactId>opentelemetry-thrift-0.13</artifactId>
    <version>OPENTELEMETRY_VERSION</version>
  </dependency>
</dependencies>
```

For Gradle, add the following to your dependencies:

```groovy
implementation("io.opentelemetry.instrumentation:opentelemetry-thrift-0.13:OPENTELEMETRY_VERSION")
```

### Usage

The instrumentation library provides the implementation of `ClientInterceptor` and `ServerInterceptor` to provide OpenTelemetry-based spans and context propagation.

```java
// For server-side, decorate processor with a tracing wrapper.
TProcessor configureServer(OpenTelemetry openTelemetry, TProcessor processor, String serviceName) {
  ThriftTelemetry thriftTelemetry = ThriftTelemetry.create(openTelemetry);

>   return telemetry.wrapServerProcessor(processor, serviceName);

Reviewer: [General] The variable is declared `thriftTelemetry` but `telemetry` is used on the next line — the snippet won't compile. Same in the two snippets below.

```suggestion
```java
// For server-side, decorate processor with a tracing wrapper.
TProcessor configureServer(OpenTelemetry openTelemetry, TProcessor processor, String serviceName) {
  ThriftTelemetry thriftTelemetry = ThriftTelemetry.create(openTelemetry);
  return thriftTelemetry.wrapServerProcessor(processor, serviceName);
```

}

// For client-side, decorate protocol with a tracing wrapper.
TProtocol configureClient(OpenTelemetry openTelemetry, TProtocol processor, String serviceName) {
  ThriftTelemetry thriftTelemetry = ThriftTelemetry.create(openTelemetry);

>   return telemetry.wrapClientProtocol(protocol, serviceName);

Reviewer: [General] Same `telemetry` typo, plus the parameter is named `processor` but typed `TProtocol` — should be `protocol` to match the call below.

```suggestion
// For client-side, decorate protocol with a tracing wrapper.
TProtocol configureClient(OpenTelemetry openTelemetry, TProtocol protocol, String serviceName) {
  ThriftTelemetry thriftTelemetry = ThriftTelemetry.create(openTelemetry);
  return thriftTelemetry.wrapClientProtocol(protocol, serviceName);
```

}

// For non-blocking client, decorate protocol factory and the async client with a tracing wrappers.
TProtocol configureClient(OpenTelemetry openTelemetry, TProtocolFactory protocolFactory, String serviceName, TTransport transport) {
  ThriftTelemetry thriftTelemetry = ThriftTelemetry.create(openTelemetry);

>   return telemetry.wrapClientProtocolFactory(protocolFactory, serviceName, transport);

Reviewer: [General] Return type should be `TProtocolFactory`, not `TProtocol` — `wrapClientProtocolFactory` returns a factory. Same `telemetry` undefined-variable issue.

```suggestion
// For non-blocking client, decorate protocol factory and the async client with a tracing wrappers.
TProtocolFactory configureClient(OpenTelemetry openTelemetry, TProtocolFactory protocolFactory, String serviceName, TTransport transport) {
  ThriftTelemetry thriftTelemetry = ThriftTelemetry.create(openTelemetry);
  return thriftTelemetry.wrapClientProtocolFactory(protocolFactory, serviceName, transport);
```

}

CustomService.AsyncIface configure(CustomService.AsyncClient asyncClient) {

>   return telemetry.wrapAsyncClient(asyncClient, CustomService.AsyncIface.class);

Reviewer: [General] Same `telemetry` undefined-variable issue. The function also has no `OpenTelemetry`/`ThriftTelemetry` parameter or local — pick one form consistent with the snippets above.

```suggestion
CustomService.AsyncIface configure(ThriftTelemetry thriftTelemetry, CustomService.AsyncClient asyncClient) {
  return thriftTelemetry.wrapAsyncClient(asyncClient, CustomService.AsyncIface.class);
```

  return thriftTelemetry.wrapServerProcessor(processor, serviceName);
}

// For client-side, decorate protocol and client with a tracing wrappers.
TProtocol configureClient(OpenTelemetry openTelemetry, TProtocol protocol) {
  ThriftTelemetry thriftTelemetry = ThriftTelemetry.create(openTelemetry);
  return thriftTelemetry.wrapClientProtocol(protocol);
}

CustomService.Iface configureClient(CustomService.Client client) {
  ThriftTelemetry thriftTelemetry = ThriftTelemetry.create(openTelemetry);
  return thriftTelemetry.wrapClient(client, CustomService.Iface.class);
}

// For non-blocking client, decorate protocol factory and the async client with a tracing wrappers.
TProtocolFactory configureClient(OpenTelemetry openTelemetry, TProtocolFactory protocolFactory) {
  ThriftTelemetry thriftTelemetry = ThriftTelemetry.create(openTelemetry);
  return thriftTelemetry.wrapClientProtocolFactory(protocolFactory);
}

CustomService.AsyncIface configureClient(CustomService.AsyncClient asyncClient) {
  ThriftTelemetry thriftTelemetry = ThriftTelemetry.create(openTelemetry);
  return thriftTelemetry.wrapAsyncClient(asyncClient, CustomService.AsyncIface.class);

> }

Reviewer: Fixed in f4e6b67e0837bf70b31479ba4f34f674fa466c11.
